tree_unique_args/src/type_analysis.egg

```diff
@@ -0,0 +1,231 @@
+      ((HasTypeDemand x))
+      :ruleset type-analysis)
+(rule ((HasTypeDemand (Get e idx)))
+      ((HasTypeDemand e))
+      :ruleset type-analysis)
+
+; Pure Op Compute
+(rule (
+       (HasTypeDemand (Add x y))
+       (HasType x (IntT))
+       (HasType y (IntT))
+      )
+      (
+       (HasType (Add x y) (IntT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (Sub x y))
+       (HasType x (IntT))
+       (HasType y (IntT))
+      )
+      (
+       (HasType (Sub x y) (IntT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (Mul x y))
+       (HasType x (IntT))
+       (HasType y (IntT))
+      )
+      (
+       (HasType (Mul x y) (IntT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (LessThan x y))
+       (HasType x (IntT))
+       (HasType y (IntT))
+      )
+      (
+       (HasType (LessThan x y) (BoolT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (And x y))
+       (HasType x (BoolT))
+       (HasType y (BoolT))
+      )
+      (
+       (HasType (And x y) (BoolT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (Or x y))
+       (HasType x (BoolT))
+       (HasType y (BoolT))
+      )
+      (
+       (HasType (Or x y) (BoolT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (Not x))
+       (HasType x (BoolT))
+      )
+      (
+       (HasType (Not x) (BoolT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (Get e n))
+       (HasType e (TupleT tylist))
+      )
+      ((HasType (Get e n) (TypeList-ith tylist n)))
+      :ruleset type-analysis)
+
+; Effectful Ops
+(rule ((HasTypeDemand (Print e)))
+      ((HasType (Print e) (UnitT)))
+      :ruleset type-analysis)
+(rule ((HasTypeDemand (Write addr val)))
+      ((HasType (Write addr val) (UnitT)))
+      :ruleset type-analysis)
+(rule ((HasTypeDemand (Read addr)))
+      ((HasType (Read addr) (IntT))) ; all reads are ints
+      :ruleset type-analysis)
+
+; Switch 
+; peel off a branch and demand type
+(rule ((HasTypeDemand (Switch cond (Cons branch rest))))
+      (
+       (HasTypeDemand branch)
+       (HasTypeDemand (Switch cond rest))
+      )
+      :ruleset type-analysis)
+; base case- demand the type of the condition
+(rule ((HasTypeDemand (Switch cond Nil)))
+      ((HasTypeDemand cond))
+      :ruleset type-analysis)
+
+; base case- single branch switch has type of branch
+(rule (
+       (HasTypeDemand (Switch cond (Cons branch (Nil))))
+       (HasType cond BoolT)
+       (HasType branch ty)
+      )
+      ((HasType (Switch cond (Cons branch (Nil))) ty))
+      :ruleset type-analysis)
+; recursive case
+(rule (
+       (HasTypeDemand (Switch cond (Cons branch rest)))
+       (HasType (Switch cond rest) ty)
+       (HasType branch ty)
+      )
+      ((HasType (Switch cond (Cons branch rest)) ty))
+      :ruleset type-analysis)
+
+; Sequencing
+(rule ((HasTypeDemand (All ord (Cons hd tl))))
```

Comment:
Optional: I would put all the demand rules together in the file, then all the typechecking rules.